### PR TITLE
Ft api social login 167313401

### PR DESCRIPTION
**Pull request by Anguandia**
#### What does this PR do?

  Add the social login feature

#### Description of Task to be completed?

  - Login from social account if user exists
  - Create user and login if social user does not exist

#### How should this be manually tested?

  - If in a local environment, clone and install this repo locally and start the server with nom start/run dev common
  - In a browser enter the url http://:base-url/login/:social-account
    - The base-url is localhost:3000 for local testing and https://codepirates-ah-backend.herokuapp.com/ for remote testing
    - Social-account is facebook or google or twitter for the respective logins
  - In the resulting window; [facebook](https://user-images.githubusercontent.com/36167006/62922655-3869e780-bdac-11e9-8931-ee307bc5c380.png), [google](https://user-images.githubusercontent.com/36167006/62927228-6522fc80-bdb6-11e9-91a1-b5d6ff52d046.png), [twitter](https://user-images.githubusercontent.com/36167006/62923881-102fb800-bdaf-11e9-81c4-ef976d9d132d.png), enter your social media account credentials
  - If the account does not exist on author's haven, you will receive a [response](https://user-images.githubusercontent.com/36167006/62924005-74eb1280-bdaf-11e9-9bb4-f6491b6812eb.png) to consent to creating an account with Author's Haven
    - Enter localhost:3000/signup/social in the address bar and hit enter, you will see the successful signup message in the [response page](https://user-images.githubusercontent.com/36167006/62924131-b7acea80-bdaf-11e9-98e2-9a1b8b593412.png)
    - Open your associated email and confirm the address
    - Continue with theses steps
  - If the user account already exists on Author's Haven, you will  receive the [login success](https://user-images.githubusercontent.com/36167006/62922821-a57d7d00-bdac-11e9-891a-77a3678b9877.png) response containing your particulars and a token

#### What are the relevant pivotal tracker stories?

  [finishes #167313401](https://www.pivotaltracker.com/story/show/167313401)

#### Screenshots (if appropriate)

facebook login page after entering localhost:3000/login/facebook in browser address bar in incognito window

facebook login page on virgin browser

Alternative on personalised browser

possible facebook interface on given browser

google login window on incognito window after entering localhost:3000/login/google in browser address bar

Screen Shot 2019-08-13 at 10 36 27

Google login on personalised browser after entering localhost:3000/login/google in browser address bar

google login

Twitter login window after entering localhost:3000/login/twitter in browser address bar

Screen Shot 2019-08-13 at 09 43 55

Response prompting non existing social user consent for signup

Screen Shot 2019-08-13 at 09 47 30

Response window on successful signup of social user

Screen Shot 2019-08-13 at 09 49 37

successful login response

Successful social login response

#### Questions:

### CODE REVIEW

File: src/controllers/socialLogin.js
Comment (habinezadalvan) [inline src/controllers/socialLogin.js]: @Anguandia So far you are on good progress. However, when a user is using social login and the user is not in Authors' heaven database he immediately signs up without making sure that he wants to sign up to Authors' heaven. I think it could be better if we first ask him if he is willing to proceed with sign up.

**Comment by Mnickii** [comment]
Would you add screenshots to this? Is it still Work In Progress?

File: src/controllers/social.js
Comment (minega25) [inline src/controllers/social.js]: I think unused code should be removed

File: src/routes/api/index.route.js
Comment (minega25) [inline src/routes/api/index.route.js]: same thing here as well. Remove unused code.

File: src/routes/index.js
Comment (minega25) [inline src/routes/index.js:39]: In my opinion, I think we should have our `api base url` before all routes in our API

**Review by habinezadalvan — CHANGES_REQUESTED** [review]
@Anguandia  I love the work you did the whole weekend, it seems you didn't sleep. Again I appreciate the fact that you worked on the feedback I gave you last time about asking the user if he wants to sign up to the Authors' heaven in case he doesn't have an account. However, as I was reviewing your work I realized that I can not log in with facebook; I think you have to give everyone access on facebook.
Check the screenshot below. Otherwise, you have my approval once that is done. Thank you.

Screen Shot 2019-08-13 at 08 02 46

File: test/users.test.js
Comment (minega25) [inline test/users.test.js]: You can also remove this code as well.

**Review by minega25 — CHANGES_REQUESTED** [review]
I think this is a job well done. Apart from some minor changes to make our codebase adhere to standards

**Review by habinezadalvan — COMMENTED** [review]
@Could you please add more descriptions on how this feature can be tested locally. Please give the Url(s) for login and sign up in case the user doesn't exist on every social login. Eg: http://localhost:3000/login/facebook. Just specify every URL you think is important to us. Otherwise, I appreciate your work. Thank you.

**Review by nshutijonathan — COMMENTED** [review]
Good work so far, however, the tests are not passing online We need to fix this

**Review by nkalyesubula — CHANGES_REQUESTED** [review]
It looks good to me. The PR is failing because this repo on your side, is still tested under travis.org. Please switch to the travis-ci platform and push again the changes. Also, address the previous comments i.e add screenshots for twitter login, and remove comments.

File: src/controllers/socialLogin.js
Comment (Anguandia) [reply src/controllers/socialLogin.js]: Factored in, thanks for the observation

File: test/verify.email.test.js
Comment (Anguandia) [reply test/verify.email.test.js]: Thanks for the observation, implemented

File: src/controllers/social.js
Comment (Anguandia) [reply src/controllers/social.js]: Thanks Patrick, sorry forgot to remove commented-out code, let me work on these right away

File: src/routes/index.js
Comment (Anguandia) [reply src/routes/index.js:39]: Let me harmonize

**Review by salviosage — COMMENTED** [review]
Nice work Mike, LGTM if the previous comment issue is solved

File: .coveralls.yml
Comment (Mnickii) [inline .coveralls.yml]: This is not part of your work. This is already merged. Please rebase using the current state of develop.

File: app.json
Comment (Mnickii) [inline app.json]: please rebase this branch well, looks like some history is messed up.

**Comment by Anguandia** [comment]
@habinezadalvan the facebook app account is still a development account I am upgrading to production so everyone has access

**Review by Mnickii — CHANGES_REQUESTED** [review]
Great work Mike. 

Please make sure this PR is an accurate depiction of only the changes you have worked on. You can achieve this by correctly rebasing the up to date develop branch.

**Review by habinezadalvan — APPROVED** [review]
Great work @Anguandia;

**Review by Mnickii — APPROVED** [review]
LGTM

**Review by minega25 — APPROVED** [review]

**Comment by salviosage** [comment]
LGTM

**Review by salviosage — APPROVED** [review]

**Review by nkalyesubula — APPROVED** [review]

**Pull request merged**
